src/printer.js

```diff
@@ -339,18 +339,22 @@ function genericPrintNoParens(path, options, print, args) {
       }
 
       // Avoid indenting sub-expressions in assignment/return/etc statements.
-      if (
-        parent.type === "AssignmentExpression" ||
-        parent.type === "VariableDeclarator" ||
-        shouldInlineLogicalExpression(n) ||
+      const willBreakRegardless =
         parent.type === "ReturnStatement" ||
         (parent.type === "JSXExpressionContainer" &&
           parentParent.type === "JSXAttribute") ||
         (n === parent.body && parent.type === "ArrowFunctionExpression") ||
         (n !== parent.body && parent.type === "ForStatement") ||
-        parent.type === "ObjectProperty" ||
-        parent.type === "Property" ||
-        parent.type === "ConditionalExpression"
+        parent.type === "ConditionalExpression";
+
+      if (
+        willBreakRegardless ||
+        (shouldInlineLogicalExpression(n) && n.left.type !== n.type) ||
+        (!shouldInlineLogicalExpression(n) &&
+          (parent.type === "AssignmentExpression" ||
+            parent.type === "VariableDeclarator" ||
+            parent.type === "ObjectProperty" ||
+            parent.type === "Property"))
       ) {
```

Comment:
Maybe it would be clearer to if written like:

```js
const willBreak =
  parent.type === "ReturnStatement" ||
  (parent.type === "JSXExpressionContainer" &&
    parentParent.type === "JSXAttribute") ||
  (n === parent.body && parent.type === "ArrowFunctionExpression") ||
  (n !== parent.body && parent.type === "ForStatement") ||
  parent.type === "ConditionalExpression";

const wontBreak =
  parent.type === "AssignmentExpression" ||
  parent.type === "VariableDeclarator" ||
  parent.type === "ObjectProperty" ||
  parent.type === "Property";

const shouldGroup =
  willBreak ||
  (shouldInlineLogicalExpression(n) ? n.left.type !== n.type : wontBreak);

if (shouldGroup) {
  return group(concat(parts));
}
```

 Not sure about the naming though =P

Reply:
I think what you put in your comment is more readable, so let's update the code to use it, instead.

What do you think of renaming `shouldGroup` to `shouldNotIndent`?

Also, I think we should change `n.type` in your `shouldGroup` declaration to `"LogicalExpression"`. These are equivalent, but using the literal is more clear in my opinion.

I also don't feel like `wontBreak` and `willBreak` are the right names, but I don't quite understand the difference between them. Could you elaborate on that a bit?

Reply:
> What do you think of renaming shouldGroup to shouldNotIndent?
>
> Also, I think we should change n.type in your shouldGroup declaration to "LogicalExpression". These are equivalent, but using the literal is more clear in my opinion.

Looks good to me

> I also don't feel like wontBreak and willBreak are the right names, but I don't quite understand the difference between them. Could you elaborate on that a bit?

I'm not sure what the names should be either. What I tried to mean is that the `willBreak` cases don't need to be indented regardless, e.g.:

```js
return (
  condition &&
  condition
);
// "condition &&" will be in a new line regardless of having a inlined logical expression
return (
  condition && [
    element
  ]
);
```

The `wontBreak` cases are the ones where it won't "break" to new lines if they should inline the logical expression, where otherwise it would:

```js
const var =
  foo &&
  bar;
// and when inlining it won't break anymore
const var = foo && [
  element
];```

Reply:
Naming is hard 😅 so I'll happily accept naming suggestions :)